feat: user 도메인 테이블 구현

best11gh opened the pull request:
- User, JoinRoute, ProfileImage의 Entity 코드 작성



close #11

best11gh committed:
feat: user 도메인 테이블 구현

kyleidea1 commented on src/main/java/org/glue/glue_be/user/entity/User.java:
```diff
@@ -1,4 +1,73 @@
+
+    @NotBlank
+    @Size(max = 36)
+    @Column(name = "uuid", nullable = false, unique = true, length = 36)
```
unique key 설정한 건 좋은 것 같아요
그런데 복합 unique 제약조건을 추가하거나, 제약조건을 좀 더 명확하게 DB 수준에서 관리하고 싶다면 다음과 같이 Table 어노테이션에 uniqueConstraints를 사용하는 것도 고려해봅시다.

@Table(name = "user", uniqueConstraints = {
    @UniqueConstraint(columnNames = {"uuid"}),
    @UniqueConstraint(columnNames = {"oauth_id"}),
    @UniqueConstraint(columnNames = {"user_name"}),
    @UniqueConstraint(columnNames = {"nickname"})
})

best11gh replied:
이 리뷰를 통해 `uniqueConstraints`에 대해 처음 알게 되었고, 더 자세히 찾아보게 되었습니다.

uniqueConstraints를 사용하면 여러 컬럼의 조합에 대해 유일 제약 조건을 설정할 수 있습니다. 예를 들어, @UniqueConstraint(columnNames = {"uuid", "oauth_id"})와 같이 선언하면, "uuid"와 "oauth_id"의 값이 모두 동일한 경우에만 제약 조건이 위반되어 오류가 발생하게 됩니다. 이는 여러 컬럼의 조합으로 유일성을 보장하고자 할 때 유용합니다.

하지만 저희의 경우, 일부 컬럼은 개별적으로 중복 없이 값이 유일해야 하므로, unique 속성을 사용하는 것이 간단하고 적합하다고 판단하였습니다. 또한, 각 필드 바로 위에 `unique` 속성을 명시해두면 코드상에서 해당 제약 조건을 한눈에 파악할 수 있어 더욱 명확합니다.

kyleidea1 commented on src/main/java/org/glue/glue_be/user/entity/User.java:
```diff
@@ -1,4 +1,73 @@
+    @Column(name = "certified", nullable = false)
+    private Integer certified = 0;
+
+    protected User() {
```
JPA가 내부적으로 리플렉션을 이용해서 객체를 생성하기에 기본 생성자를 만들어 주고, 이 때 JPA만 접근하기 위해 접근제한자로는 protected가 권장된다고 합니다.
잘 생성해주셨네요!

kyleidea1 commented on src/main/java/org/glue/glue_be/user/entity/User.java:
```diff
@@ -1,4 +1,73 @@
+    protected User() {
+    }
+
+    public User(String uuid, String oauthId, String userName, String nickname, Integer gender, LocalDate birth,
```
AllArgsConstructor를 안 쓰고 일일이 작성한 부분이 학습적인 측면에서 좋다고 판단되네요.
Lombok을 얼만큼 사용할 지 추후 팀원들과 논의해봅시다. 장단점이 있으니까

best11gh replied:
🚀 논의 완료

kyleidea1 commented on build.gradle:
```diff
@@ -29,7 +29,8 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-web'
 	implementation 'org.xerial:sqlite-jdbc:3.41.2.1' // sqlite jdbc
```
jpa - jdbc 관계를 한 번 훑어보면 좋을 것 같아요!

kyleidea1 commented on src/main/java/org/glue/glue_be/user/entity/JoinRoute.java:
```diff
@@ -0,0 +1,31 @@
+    private User user;
+
+    @NotNull
+    @Column(name = "join_route_type", nullable = false)
```
nullability에 대해서도 학습할 것이 많은데요

우선 @NotNull 어노테이션은 jakarta.validation.constraints 소속으로 애플리케이션 레벨에서 동작합니다. 그리고 Notnull은 Spring Validator/ Hibernate Validator 등에서 @Valid 어노테이션을 통해 작동시킵니다. 이 @Notnull 어노테이션 자체로는 아무 일도 하지 않고, '이거 null이면 안 돼'하고 Validator에게 알려주는 표식같은 거에요.

@Column의 nullable = false는 jakarta.persistence 소속으로 DB 레벨에서 동작하고, JPA가 DDL 생성 시 해당 DB에 제약 조건으로써 추가하는 개념이에요. 애초에 null값이 데이터베이스에 저장되는 걸 차단하는 부분입니다.

결론적으로, NotNull - 앱 레벨, nullable = false - DB 레벨 에서 이중으로 null 값이 저장되는 것을 차단하도록 잘 작성됐네요!

best11gh replied:
🚀 논의 완료

kyleidea1 commented on src/main/java/org/glue/glue_be/user/entity/JoinRoute.java:
```diff
@@ -0,0 +1,31 @@
+    private Long joinRouteId;
+
+    @OneToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "user_id", nullable = false, unique = true)
```
OneToOne 매핑에서 주의할 점을 좀 적어볼게요

우선 optional 파라미터는 기본값이 true이기 때문에, optional = false가 명시되지 않으면 이 연관 테이블이 null이어도 허용됩니다. join_route는 저희가 회원가입 UI에서 필수로 받아오기 때문에 nullable 하지 않으므로, 잘 작성된 것 같아요.

그리고 JoinColumn 어노테이션을 통해 반드시!! 외래키를 명시적으로 적어줘야 합니다. 그렇지 않으면 DB 컬럼명과 제약 조건 등을 자동으로 추정해서 작성해버려서 유지보수성이 확 떨어지게 되구요. 잘 하셨어요

@OneToOne은 그 자체로서 1:1 연관관계 매핑을 DB 수준에서 보장하지 않습니다. 그래서 JoinColumn의 unique = true를 반드시 기입해줘야, 비로소 1:1 연관관계 매핑이 됩니다. 잘 작성하셨네요

FetchType에 대해서는 다룰 내용이 너무 많아서 다음 스터디 때 다 같이 얘기해보는 걸로 합시다!

kyleidea1 replied:
또한, 이 일대일 매핑이 단방향인지, 양방향인지에 따라 연관관계의 주인을 명시해야 합니다. 현재는 JoinRoute 테이블만이 User 테이블로부터 외래키를 받아오는 구조이기 때문에 단방향 일대일 매핑이라 잘 작성해주셨습니다.
다만, 양방향으로의 확장을 고려한다면 mappedBy를 이용하여 연관관계의 주인을 잘 정의해 주셔야 합니다.
시간이 나면 단뱡향 1:1 매핑을 서로 하는 것과 양방향 1:1 매핑이 어떻게 다른 지 학습해보시면 좋을 것 같습니다.

best11gh replied:
🚀 논의 완료

best11gh committed:
refactor(entity): 엔티티 빌더 적용

- User, ProfileImage, JoinRoute 엔티티에 Lombok @NoArgsConstructor와 @Builder 적용

best11gh committed:
refactor: 엔티티에서 Bean Validation 제거

- DTO에서 검증하고, 엔티티는 DB 스키마 생성에만 집중하도록 역할 분리

best11gh committed:
refactor(User): 필드 변경 메서드 추가

- changeName, changeNickname, changeNation, changeDescription, changeCertified 메서드 추가

best11gh committed:
feat(profile_image): 프로필 이미지 URL 변경 메서드 추가

- changeProfileImageUrl 메서드 추가하여 프로필 이미지 URL 업데이트 가능
